### ISSUE TITLE
Creates RR, but can not find them.

### ISSUE THREAD

**Issue by twhitaker0808** [EfficientIP-Labs/terraform-provider-solidserver#26]
Running into two issues, first is when specifying a DNS view RRs are not able to create.  Second after blocking out the DNS view I am able to create the RR, but a terraform destroy is not able to find the RR and thinks it is already destroyed.

resource "solidserver_dns_rr" "helloworld" {
  dnsserver = "dd-smart-dns-internal.mydns.com"
  #dnsview   = "Main"
  dnszone   = "mydns.com"
  name      = "test0.mydns.com"
  type      = "A"
  value     = "1.2.3.4"
}

**Comment by alexissavin** [comment]
Hello,

I'm quite surprise about the problem you describe which is part of the default test suite.
The following example works pretty fine on all supported SOLIDserver versions.

`
# Configure the SOLIDserver Provider
provider "solidserver" {
  host      = var.solidserver_host
  username  = var.solidserver_user
  password  = var.solidserver_password
  sslverify = false
}

resource "solidserver_dns_smart" "myFirstDnsSMART" {
  name       = "myfirstdnssmart.priv"
  arch       = "multimaster"
  comment    = "My First DNS SMART Autmatically created"
  recursion  = true
  forward    = "none"
}

resource "solidserver_dns_server" "myFirstDnsServer" {
  name       = "myfirstdnsserver.priv"
  address    = "127.0.0.1"
  login      = "admin"
  password   = "admin"
  smart      = solidserver_dns_smart.myFirstDnsSMART.name
  smart_role = "master"
  comment    = "My First DNS Server Autmatically created"
}

resource "solidserver_dns_view" "myFirstDnsView" {
  depends_on      = [solidserver_dns_smart.myFirstDnsSMART]
  name            = "myfirstdnsview"
  dnsserver       = solidserver_dns_smart.myFirstDnsSMART.name
  recursion       = true
  forward         = "first"
  forwarders      = ["172.16.0.42", "172.16.0.43"]
  allow_transfer  = ["172.16.0.0/12", "192.168.0.0/24"]
  allow_query     = ["172.16.0.0/12", "192.168.0.0/24"]
  allow_recursion = ["172.16.0.0/12", "192.168.0.0/24"]
  match_clients   = ["172.16.0.0/12", "192.168.0.0/24"]
  match_to        = ["192.168.1.1/32"]
}

resource "solidserver_dns_view" "mySecondDnsView" {
  depends_on      = [solidserver_dns_server.myFirstDnsServer]
  name            = "mySecondDnsView"
  dnsserver       = solidserver_dns_smart.myFirstDnsSMART.name
  recursion       = true
  forward         = "none"
  forwarders      = []
  allow_transfer  = ["10.0.0.0/8"]
  allow_query     = ["10.0.0.0/8"]
  allow_recursion = ["10.0.0.0/8"]
  match_clients   = ["10.0.0.0/8"]
}

resource "solidserver_dns_zone" "myFirstZone" {
  depends_on      = [solidserver_dns_view.myFirstDnsView]
  dnsserver       = solidserver_dns_smart.myFirstDnsSMART.name
  dnsview         = solidserver_dns_view.myFirstDnsView.name
  name            = "mycompany.priv"
  type            = "master"
  notify          = "explicit"
  also_notify     = ["127.0.0.2:53", "127.0.0.3:53"]
  createptr       = false
}

resource "solidserver_dns_zone" "mySecondZone" {
  depends_on      = [solidserver_dns_view.mySecondDnsView]
  dnsserver       = solidserver_dns_smart.myFirstDnsSMART.name
  dnsview         = solidserver_dns_view.mySecondDnsView.name
  name            = "mysubcompany.priv"
  type            = "master"
  createptr       = false
}

resource "solidserver_dns_forward_zone" "myFirstForwardZone" {
  depends_on  = [solidserver_dns_view.myFirstDnsView]
  dnsserver   = solidserver_dns_smart.myFirstDnsSMART.name
  dnsview     = solidserver_dns_view.myFirstDnsView.name
  name        = "fwd.mycompany.priv"
  forward     = "first"
  forwarders  = ["10.10.8.8", "10.10.4.4"]
}

resource "solidserver_dns_rr" "AFirstRecords" {
  depends_on   = [solidserver_dns_zone.myFirstZone]
  dnsserver    = solidserver_dns_smart.myFirstDnsSMART.name
  dnsview      = solidserver_dns_view.myFirstDnsView.name
  name         = "aarecord-${count.index}.mycompany.priv"
  type         = "A"
  value        = "127.0.0.1"
  count        = 16
  class  = "CUSTOM_RR"
  class_parameters = {
    custom = "4267987"
  }
}

resource "solidserver_dns_rr" "CnameFirstRecords" {
  depends_on   = [solidserver_dns_rr.AFirstRecords]
  dnsserver    = solidserver_dns_smart.myFirstDnsSMART.name
  dnsview      = solidserver_dns_view.myFirstDnsView.name
  name         = "cnamerecord-${count.index}.mycompany.priv"
  type         = "CNAME"
  value        = "aarecord-${count.index}.mycompany.priv"
  count        = 16
}

resource "solidserver_dns_rr" "ASecondRecords" {
  depends_on   = [solidserver_dns_zone.mySecondZone]
  dnsserver    = solidserver_dns_smart.myFirstDnsSMART.name
  dnsview      = solidserver_dns_view.mySecondDnsView.name
  name         = "aarecord-${count.index}.mysubcompany.priv"
  type         = "A"
  value        = "127.0.0.1"
  count        = 16
}

resource "solidserver_dns_rr" "CnameSecondRecords" {
  depends_on   = [solidserver_dns_rr.ASecondRecords]
  dnsserver    = solidserver_dns_smart.myFirstDnsSMART.name
  dnsview      = solidserver_dns_view.mySecondDnsView.name
  name         = "cnamerecord-${count.index}.mysubcompany.priv"
  type         = "CNAME"
  value        = "aarecord-${count.index}.mysubcompany.priv"
  count        = 16
}
`

Please ensure you have proper permissions on the servers, views and zones.

If you need further assistance, please share the debug logs (See the [Readme](https://github.com/EfficientIP-Labs/terraform-provider-solidserver#debug)).

Kind regards

**Comment by twhitaker0808** [comment]
[plan.log](https://github.com/EfficientIP-Labs/terraform-provider-solidserver/files/12487013/plan.log)
[destroy.log](https://github.com/EfficientIP-Labs/terraform-provider-solidserver/files/12487014/destroy.log)
[apply.log](https://github.com/EfficientIP-Labs/terraform-provider-solidserver/files/12487015/apply.log)


Attached plan, apply and destroy logging

**Comment by alexissavin** [comment]
Hello,

In the provided logs, I see no reference to any view:

`  # solidserver_dns_rr.helloworld will be created
  + resource "solidserver_dns_rr" "helloworld" {
      + dnsserver = "dd-smart-dns-internal.intra.renhsc.com"
      + dnszone   = "intra.renhsc.com"
      + id        = (known after apply)
      + name      = "myazuretst.intra.renhsc.com"
      + ttl       = 3600
      + type      = "A"
      + value     = "8.8.8.8"
    }`

You need to specify in which view the Zone and  RR are located.

I suppose I should prevent the default behavior that allow the creation of this RR in the default view transparently (I will investigate this point).

Kind regards.

**Comment by twhitaker0808** [comment]
Correct, When specifying the view.  The plan fails to create the record with having an issue with the DNS zone.

solidserver_dns_rr.helloworld: Creating...
╷
│ Error: Unable to create RR: myazuretst01.intra.renhsc.com (RR cannot belong to given DNS zone)
│
│   with solidserver_dns_rr.helloworld,
│   on a_record_test.tf line 17, in resource "solidserver_dns_rr" "helloworld":
│   17: resource "solidserver_dns_rr" "helloworld" {

Only when the view is omitted will it create the record (oddly enough in the correct view)  but then no longer able to find it when destroying.

**Comment by alexissavin** [comment]
Well this "oddly" behavior is because the SOLIDserver service does its best.
If no view is specified, it create the RR within all matching views and zones.
The delete does not do the same, terraform looks for the exact RR and delete
the exact RR based on its ID.

I think I will require the view name in the terraform resource upon creation,
if the DNS server leverage any view.

Let me know if this sound an acceptable "fix"/enhancement.

Kind regards

**Comment by twhitaker0808** [comment]
No, Still no luck.  We have opened a ticket in the support portal.  Perhaps we can schedule time to look at it together, maybe I'm missing something very obvious.  But any testing done fails so far.

**Comment by alexissavin** [comment]
Well I must say that I'm a bit lost.

The last error you shared is : "RR cannot belong to given DNS zone".
Meaning the RR doesn't belongs to the zone you specified.
It's not related to the view.

For the records, I have push a modification that now requires the view name
when creating a Zone or RR on a server that has view configured.

Hope this helps.

**Comment by twhitaker0808** [comment]
Understandable why you might be lost, I'm  just as lost :D.

None of the views/zones have been built and created with the TF provider.  I am simply attempting to create records out of band.  A records in the zone specified is and should be legitimate and the account used can create the same record locally when not using TF.

When specifying the view - failed to create records.  When not specifying creates the record however can not remove the record.

I will test again with the latest version and see if anything functions as expected.

**Comment by alexissavin** [comment]
Hello,

Did you had a chance to build and try with the current master branch ?

Thanks in advance for your feedback.

Kind regards.

**Comment by hallaj** [comment]
I'm facing similar issues, not sure if it's related to be honest.
```
data "solidserver_dns_smart" "main" {
  name = "smartdns.domain.tld"
}

data "solidserver_dns_view" "common" {
  name      = "common-view"
}

resource "solidserver_dns_zone" "my_custom_zone" {
  name      = "domain.tld"
  dnsserver = data.solidserver_dns_smart.main.name
  dnsview   = title(data.solidserver_dns_view.common.name)
  type      = "master"
}
  
resource "solidserver_dns_rr" "my_test_record" {
  dnsserver  = data.solidserver_dns_smart.main.name
  dnsview    = title(data.solidserver_dns_view.common.name)
  dnszone    = solidserver_dns_zone.my_custom_zone.name
  name       = "my_test_record.domain.tld"
  ttl        = "600"
  type       = "CNAME"
  value      = "baserecord.domain.tld"
}
```

Getting the following result:
```
solidserver_dns_rr.my_test_record: Creating...
╷
│ Error: Unable to create RR: my_test_record.domain.tld (RR cannot belong to given DNS zone)
│
│   with solidserver_dns_rr.my_test_record,
│   on main.tf line 21, in resource "solidserver_dns_rr" "my_test_record":
│   21: resource "solidserver_dns_rr" "my_test_record" {
│
╵
```

The following happens with the current `master` branch's code.

---

Not sure if it helps, but the zone was created via the web UI and I did a `terraform import` to fetch the state.

**Comment by alexissavin** [comment]
Hello Hallaj,

I'm sorry, but I hardly think this is the terraform script you use to reproduce this error
that I do not reproduce on my side. The provided terraform script is invalid in many ways.

First, your data-source is missing a parameter:
```
│ Error: Missing required argument
│ 
│   on tests.tf line 27, in data "solidserver_dns_view" "common":
│   27: data "solidserver_dns_view" "common" {
│ 
│ The argument "dnsserver" is required, but no definition was found.
```

Second, the RR name or value can not contains '_':
```
│ Error: Unable to create RR: my_test_record.domain.tld (Bad parameter format)
│ 
│   with solidserver_dns_rr.my_test_record,
│   on tests.tf line 39, in resource "solidserver_dns_rr" "my_test_record":
│   39: resource "solidserver_dns_rr" "my_test_record" {
```

Also, you must not use the 'title' function which capitalize the first letter of the string, which is unexpected.

Once fixed, I work with the following terraform script:
```
data "solidserver_dns_smart" "main" {
  name = "smart.local"
}

data "solidserver_dns_view" "common" {
  name      = "secondview"
  dnsserver = data.solidserver_dns_smart.main.name
}

resource "solidserver_dns_zone" "my_custom_zone" {
  name      = "domain.tld"
  dnsserver = data.solidserver_dns_smart.main.name
  dnsview   = data.solidserver_dns_view.common.name
  type      = "master"
}

resource "solidserver_dns_rr" "my_test_record" {
  dnsserver  = data.solidserver_dns_smart.main.name
  dnsview    = data.solidserver_dns_view.common.name
  dnszone    = solidserver_dns_zone.my_custom_zone.name
  name       = "testrecord.domain.tld"
  ttl        = "600"
  type       = "CNAME"
  value      = "baserecord.domain.tld"
}
```

I do not reproduce this issue with this one.

Kind regards

**Comment by hallaj** [comment]
hey @alexissavin,

Apologies for the bad copy pasting job I did. You're right - I accidentally left out some of the parameters when pasting.

The `title` function was actually basing off my `terraform import`. Allow me to share a bit of the background of our setup.

In our current setup, one of the views we have created (In the web ui) is called `Common-View`. When I attemped a `terraform import solidserver_dns_rr.some_record 1234`, the terraform state returned `"dnsview": "Common-View"`. Below are the snippets of it:
```
          "attributes": {
            "class": null,
            "class_parameters": null,
            "dnsserver": "smartdns.domain.tld",
            "dnsview": "Common-View",
            "dnszone": "domain.tld",
            "id": "1234",
            "name": "somerecord.domain.tld",
            "ttl": 600,
            "type": "A",
            "value": "1.2.3.4"
          },
```

 Prior to be using the `title` keyword, a plan would show the following, which explains why I added the `title` keyword to ensure nothing changes between the record and the current terraform state.
```
  # solidserver_dns_rr.wildcard_pages_infra_paynet_my must be replaced
-/+ resource "solidserver_dns_rr" "somerecord" {
      ~ dnsview   = "Common-View" -> "common-view" # forces replacement
      ~ id        = "1234" -> (known after apply)
        name      = "somerecord.domain.tld"
        # (5 unchanged attributes hidden)
    }
```

---

I found something interesting after looking at the `terraform.tfstate`. The `datasource` for the view did return `common-view` instead of `Common-View`, however the `terraform import` for an existing `solidserver_dns_rr` returned `Common-View` instead of `common-view`.

---

Just to add, strangely enough, even without the `title` keyword, I am still receiving the same error `Error: Unable to create RR: mytestrecord.domain.tld (RR cannot belong to given DNS zone)`

**Comment by alexissavin** [comment]
Sounds like a case sensitivity issue :/

Please give a try to the branch "https://github.com/EfficientIP-Labs/terraform-provider-solidserver/tree/fix%2326".

It seems that the view name is actually case sensitive.

**Comment by hallaj** [comment]
Thanks @alexissavin - that addressed the record creation bit. 

Record deletion and update (this performs a record replacement) doesn't seem to work.

```
Error: Unable to delete RR: somerecord.domain.tld
```

**Comment by hallaj** [comment]
[tf-debug.apply.log](https://github.com/EfficientIP-Labs/terraform-provider-solidserver/files/13235208/tf-debug.apply.log)
[tf-trace.apply.log](https://github.com/EfficientIP-Labs/terraform-provider-solidserver/files/13235209/tf-trace.apply.log)

hi @alexissavin - Attaching both debug and trace logs when I perform `terraform apply -auto-approve`.

**Comment by alexissavin** [comment]
@hallaj , do you have the terraform file ?
I don't understand how you get the following values: "newrecord**..**domain.tld", "traefik..domain.tld".
This are not valid ones...

**Comment by hallaj** [comment]
[logs.zip](https://github.com/EfficientIP-Labs/terraform-provider-solidserver/files/13238148/logs.zip)

Attaching the terraform file and logs.

**Comment by alexissavin** [comment]
Hello,

Please give a try to version v1.1.20, this should fix your issue.

Kind regards

**Comment by twhitaker0808** [comment]
v1.1.20 fixes the issues where destroy was not able to find the records.

**Comment by alexissavin** [comment]
@twhitaker0808 thanks for your patience and for your help.

**Comment by twhitaker0808** [comment]
Do you have the root cause of why that was behaving in that manner?

**Comment by alexissavin** [comment]
Yes, DNSviews are actually case sensitive in SOLIDsever. This provider did not implemented this behavior. The views were handled just as DNS server names (lower case).
For convenience, view names were always converted to lower case, so in the end, the created views were always in lower case. Initial attempts to fix this did not addressed the fact that the previously created views (v <= 1.1.19) were lower case.

**Comment by twhitaker0808** [comment]
So in theory if I passed the view in all lower case, <=1.1.19 it should have worked.  Vs as created on the actual appliance?

**Comment by alexissavin** [comment]
Yes, so far that I understood and tested, there was no issue with lower case views.
